### ISSUE TITLE
[RA] Upgrade observer/replay interface features.

### ISSUE THREAD

**Issue by SoScared** [OpenRA/OpenRA#5121]
EDIT: Narrowing the aim down to the RA mod exclusively. If one day successful it could leak on to the other mods. Would be worth having for the current ORA flagship.

I'd love to see a re-design of the statistics bar as suggested in #4716 for all default mods (TD, RA & D2K) with hotkeys bound to each tab.

Due to the long discussion at #4716 and recent interest I'll clarify the goals on this issue (details below) for anyone wanting to take a stab at it.

**Comment by SoScared** [comment]
It's been a good year. Adding bounty to my favourite feature request. Good hunting!

**Comment by SoScared** [comment]
Due to the long discussion at #4716 and more attention on the subject, I just want to clarify the goals of this issue for anyone wanting to take a stab at it:

Newly opened discussion at: http://www.sleipnirstuff.com/forum/viewtopic.php?f=82&t=18631

Make the interface resemble some of the mockups from #4716  with information popping up at the top right of the screen, and allow access to the different tabs (production, economy etc.) with hotkeys. The initial mockups looks the most interesting:

![68747470733a2f2f662e636c6f75642e6769746875622e636f6d2f6173736574732f363639323136372f323234303636372f34316463643862632d396361322d313165332d383234382d3436636434643166656664392e6a7067](https://cloud.githubusercontent.com/assets/2715583/10264433/b8f9b4aa-6a0b-11e5-8780-3626fd696833.jpg)
![68747470733a2f2f662e636c6f75642e6769746875622e636f6d2f6173736574732f3735363636392f323336353230392f64316333663133652d613662332d313165332d393566622d6439313738346432633335662e706e67](https://cloud.githubusercontent.com/assets/2715583/10264458/4d3c44b6-6a0c-11e5-80ca-951f0bfa3b9a.png)
![68747470733a2f2f662e636c6f75642e6769746875622e636f6d2f6173736574732f3735363636392f323336353230322f37363266363530362d613662332d313165332d386632302d6335616263353831373936632e706e67](https://cloud.githubusercontent.com/assets/2715583/10264459/4fe75eee-6a0c-11e5-915c-665ad33789a9.png)

For me the suggested production tab above is the meat of the matter. It just looks so good. Matches with more than 8 players might make the prod.tab a bit huge, but what of it. 10-player games aren't those you normally follow close on production anyways. The other tabs will do just fine as long as they look pretty and with transparent panels as suggested in the above mockups.

**Comment by SoScared** [comment]
So I've been carefully studying the extensive production capabilities in Tiberian Dawn in a series of long, heavy macro team-games. The number of unique buildings/units produced per player at any given time rarely exceeds 10 and it seems to never pass 13 in any match. Tiberian Dawn has the potential of making up to 32 unique buildings/units (Nod) at a time which would require 9 MCV's 5 Barracks, 9 Airstrips, 2 Helipads, all producing one unique item simultaneously. That just won't happen.

I've looked into the same issue for Starcraft 2, which has been the spark of inspiration for this issue. The same principle applies there. Despite having the capability to producing over 45 (!) unique buildings/units/upgrades simultaneously, the production tab never get's out of hand, even in long macro games between in professionals in leagues and tournaments.

Mockup of a production tab with 15 80% sized slots on 1280x720p resolution. Mockup is from D2K but used as an example for Tiberian Dawn:

![80percent720_15](https://cloud.githubusercontent.com/assets/2715583/10699106/c7375186-79b4-11e5-80f9-7c6448c66533.png)

15 80% sized slots on 1920x1080p resolution:

![80percent_15](https://cloud.githubusercontent.com/assets/2715583/10699231/a1bb1d4c-79b5-11e5-8d0e-391dc8c8c46f.png)

Tiberian Dawn needs only a number indicator in it's slots (as in Starcraft 2) as seen in the above mockup for buildings/units produced at quantities more than 1.

**Comment by pchote** [comment]
If the same item is being built from multiple factory and we show a "2" overlay, do we pick the one with the least remaining time for the clock?

**Comment by SoScared** [comment]
That would be preferable. At least that's what I'm used to with SC2.

**Comment by GraionDilach** [comment]
Overlapping sounds bad. You can't use observing then for AI debugging because some mods might want to debug factory output - modders might want to use factories with overlapping output (think Warzone2100 where factory level decided the bodies the factory can produce) because the order gets really messy.

An option could be overlapping clocks with each other - since all of our clocks are transparent, more of them could be placed on top of each other while all could be still tracked.

**Comment by penev92** [comment]
Modders are free to make their own UI. Like I said a couple of days ago - the UI is the mod's responsibility, and that includes the observer UI.

I don't expect stacking the sprites to look good at all.

**Comment by Mailaender** [comment]
https://youtu.be/0IDixZguIP0 shows a nice live mockup.

**Comment by SoScared** [comment]
https://www.youtube.com/watch?v=ILd0DqW7qqQ a more closely resembling mockup.

**Comment by SoScared** [comment]
Narrowing down the focus towards the RA mod instead of cross-mods.

**Comment by rob-v** [comment]
I don't do graphics so I only tried to restyle the Statistics window - it has the same functionality like current window (+ tooltip of all currently produced items).

![obrazok](https://cloud.githubusercontent.com/assets/16348750/24248798/23a8bbd6-0fd1-11e7-8d05-a430fae0c1f2.png)
![obrazok](https://cloud.githubusercontent.com/assets/16348750/24248816/38aeae46-0fd1-11e7-978f-539d21451496.png)

**Comment by SoScared** [comment]
With some blur shadow/border behind the icons/information the production bar looks like a good step in the right direction. I think for the stats tho we might want to have a faint black transparent background. If possible it might look better to narrow the vertical distance between the stat rows.

**Comment by Smittytron** [comment]
Here's another request that I hope isn't too outlandish: Hotkeys for observer/replay mode.
Would be awfully convenient for streaming/casting if you can fly through the production/combat/eco tabs with hotkeys. Would also want a key to leave the stats.

**Comment by WolfGaming** [comment]
@Smittytron If you look at the topmost post you see he already asked for that "with hotkeys bound to each tab.".

**Comment by Smittytron** [comment]
My bad

**Comment by rob-v** [comment]
The shortcuts are now also working. These shortcuts should be added to Settings - Hotkeys section. There is more work. I don't know if I should continue, isn't there some better/nice work in progress?

**Comment by SoScared** [comment]
> isn't there some better/nicer work in progress?

Nope! It still says @penev92 is assigned to the issue but he said himself he has left the ticket.

**Comment by rob-v** [comment]
This design is acceptable for you? In current Statistics window you don't see full queue, only currently produced buildings/units. I would like to keep it at least for 1st phase, otherwise it could take longer, be rejected for some reasons, ...

**Comment by SoScared** [comment]
Personally I think it looks like a great improvement! Perhaps some darker color on white strip borders, but it's nitpicking. Also I can't speak to how well the design implementation correlate with the overall style and intent with the developers.

Also any chance to have the transparent background icon/text area shrink on games with fewer players (and perhaps also a bit horizontally)? Say e.g. a 2-player match stat tab looking more like this (copy/paste mock-up):
![22033266-0fea-11e7-960a-0b54a72a95ff](https://cloud.githubusercontent.com/assets/2715583/24264063/731466f6-0fff-11e7-9408-97b4c9938df0.png)
![255968c2-0fea-11e7-90bd-f2348b7fc20d](https://cloud.githubusercontent.com/assets/2715583/24264065/75ca6e86-0fff-11e7-80ab-70f2cd8547bf.png)

**Comment by rob-v** [comment]
We can wait on @pchote comment before I continue/finish it. There shouldn't be major implementation objections, all is already done. Yes, it would shrink based on number of players. Re horizontally - you mean e.g. Economy lines less space between lines or also smaller icons in Production tab?

**Comment by GraionDilach** [comment]
I'd like to see a vertical bar there - the big icons worry me with mods using individual production like TD.

**Comment by SoScared** [comment]
@rob-v just having the transparent background area as small as possible. I love the big icons and text sizes.

**Comment by SoScared** [comment]
Would be nice if evt. horizontal/vertical bar could pop-up depending on the amount of players (vertical) and production slots (horizontal).

**Comment by rob-v** [comment]
@GraionDilach  You mean horizontal bar? I didn't see any so far in OpenRA, so I don't know how easy it is to use one.
@SoScared There is already vert. scrollbar in case of more players. It should work. Would you like to keep the height always the same (e.g. for 2 players) - simpler or adjust e.g. up to 4 players + use scrollbar?

**Comment by pchote** [comment]
The layout looks good to me, but I think we'll need to come up with some more subtle panel artwork instead of reusing the tooltip version.

**Comment by rob-v** [comment]
You mean the tooltip with list of currently produced items? That one I have PR open, but it is for current Statistics window with small icons, less readable. I removed the tooltip for this version with large icons.

I agree, artwork can be much improved, but I am not able to do, so I let this task for someone else. I just experimented today what could be done with current Statistics window.

**Comment by SoScared** [comment]
If the scroll bar could auto-hide when there's nothing to scroll that would be great.

If the height (with production) would adjust itself to the amount of players, starting from 1 (missions, playing around alone etc.) and increasing up to idk, 8 players, and then make the vertical scroll-bar pop up if you exceed 8 players. That would be stellar.

And yeah, generally just keep the dead space as small as possible.

In fact, if the above gets through the devs with the requested changes on a PR I'm willing to call mission accomplished for this issue's bounty. If it's not premature I think it's worth putting an end to this issue this way.

**Comment by pchote** [comment]
> You mean the tooltip with list of currently produced items?

I mean the black background with the white rounded border, which is used for the tooltips in RA.

**Comment by SoScared** [comment]
Replacing the white border with black would immediately make the info easier on the eye.
![22033266-0fea-11e7-960a-0b54a72a95ff](https://cloud.githubusercontent.com/assets/2715583/24267966/4eb0ef34-100c-11e7-95fe-618ce9c3d402.png)
![255968c2-0fea-11e7-90bd-f2348b7fc20d](https://cloud.githubusercontent.com/assets/2715583/24267968/503eaddc-100c-11e7-996e-02d7bfd314a0.png)

Again, just simple paint mock-up. Also removed the scroll bars.

**Comment by rob-v** [comment]
@pchote It seems that we need new dialog in chrome.yaml (e.g. to avoid white border) and update ScrollPanelWidget to support hiding ScrollBar. ScrollbarWidth = 0 seems not to work. Could you please update ScrollPanelWidget to support hiding border and scrollbar? 
I could finish then resizing based on number of players, show/hide scrollbar, defining new hotkeys in Settings dialog,....

**Comment by rob-v** [comment]
I understand now better @pchotes' note re 'panel artwork'. It is defined as:
# It's a Container, used in various places, like Hotkeys, ColorPicker, Asset Browser frames (looks like a pressed button)
dialog3: dialog.png
...

# Used for tooltips and the video player
dialog4: dialog.png
	background: 517,392,54,54
	border-t: 517,387,54,6
	border-b: 517,446,54,6
	border-l: 512,392,6,54
	border-r: 571,392,6,54
	corner-tl: 512,387,6,6
	corner-tr: 571,387,6,6
	corner-bl: 512,446,6,6
	corner-br: 571,446,6,6

# completely black tile
dialog5: dialog.png
...

I used dialog4. I can create a copy for Statistics and remove white border simply here I assume. If you would like to define your own layout/colors, you could send me new dialog definition.

**Comment by rob-v** [comment]
4 players:

![obrazok](https://cloud.githubusercontent.com/assets/16348750/24290808/a4b2dcba-1086-11e7-9ad4-8342d845c0cb.png)
![obrazok](https://cloud.githubusercontent.com/assets/16348750/24290810/a98932ca-1086-11e7-94f0-076f837bdbb1.png)

12 players (with scrollbar):

![obrazok](https://cloud.githubusercontent.com/assets/16348750/24290829/c323fbde-1086-11e7-94b7-b853655a7583.png)
![obrazok](https://cloud.githubusercontent.com/assets/16348750/24290834/c7f9f640-1086-11e7-94ed-f17574fb053f.png)

I have one remaining problem after chanding width based on selected option.
![obrazok](https://cloud.githubusercontent.com/assets/16348750/24290877/fbde047e-1086-11e7-9549-0ac78923d112.png)

**Comment by MustaphaTR** [comment]
I'm pretty sure Hind is not a support power. That title may need to go a bit further.

**Comment by rob-v** [comment]
Ok, thanks. 
I was now reading another issue about this and I see now long history with great/nicer proposals, even implementations. If I read it before starting, I would never start this as it is/was already done better.
Why it was never completed? I added hotkeys, dynamic resizing, but that could be also added to previous work.

**Comment by rob-v** [comment]
TD, D2K, TS:

![obrazok](https://cloud.githubusercontent.com/assets/16348750/24307918/45befe10-10c6-11e7-964f-84f2ab2f09a4.png)
![obrazok](https://cloud.githubusercontent.com/assets/16348750/24307927/4ac83cb4-10c6-11e7-8f76-c661d8454917.png)

![obrazok](https://cloud.githubusercontent.com/assets/16348750/24307880/31ef065a-10c6-11e7-90ad-aecb2a1d58c2.png)
![obrazok](https://cloud.githubusercontent.com/assets/16348750/24307887/37096b08-10c6-11e7-9559-0d9bed966d09.png)

![obrazok](https://cloud.githubusercontent.com/assets/16348750/24307934/52610442-10c6-11e7-9085-16a7521426c3.png)
![obrazok](https://cloud.githubusercontent.com/assets/16348750/24307941/56ab6c68-10c6-11e7-976e-4096830934d0.png)

**Comment by SoScared** [comment]
Blown away.

Should mention still that with the above we might want ca. 5-6 extra pixels height on the bottom for visual sake.
![asdfsadfasdf](https://cloud.githubusercontent.com/assets/2715583/24308777/a849894e-10c9-11e7-9873-6ed907110d63.png)

**Comment by GraionDilach** [comment]
Now that I look to the TD image, I just remembered the biggest issue with that observer view which this proposal also doesn't fix: supportpowers can hide the production queues via overlapping. Having 9 production queues doesn't sound many enough for me to make sure that doesn't happen.

**Comment by MustaphaTR** [comment]
Another thing, except D2K, why Infantry Queue is after the Vehicle? I think they should go before it.

**Comment by SoScared** [comment]
> So I've been carefully studying the extensive production capabilities in Tiberian Dawn in a series of long, heavy macro team-games. The number of unique buildings/units produced per player at any given time rarely exceeds 10 and it seems to never pass 13 in any match.

Quoting comment https://github.com/OpenRA/OpenRA/issues/5121#issuecomment-150629293

The mods likely will need more custom designs regarding the default production lenght.

**Comment by rob-v** [comment]
I did TD Production column wider than others as it was overlapping in my test. This problem was then also with current window or didn't happen due to small icons? 
What do you propose - make panel wider, icons small, support powers in another tab,...?
@MustaphaTR I didn't change it, it should be like in current Statistics window, but I have to check it

**Comment by SoScared** [comment]
I for one would rather see the current design kept and look towards other solutions to the overlap problems.

**Comment by MustaphaTR** [comment]
I know, i remember it being like that on current release too, but as i said i don't think it should be like that. I would be happy if you can fix that too when you are in it. I think the issue is `ClassicProductionQueue:`'s order under `Player:` actor.

**Comment by SoScared** [comment]
How advanced would it be to stack similar production as mentioned in https://github.com/OpenRA/OpenRA/issues/5121#issuecomment-150629293 ?

**Comment by GraionDilach** [comment]
I would drop/decrease the upscaling of the cameos and place the support powers to a second row.

**Comment by rob-v** [comment]
@GraionDilach You could be right, but in general I don't want to do many 'major' UI design changes, as this isn't routine for me so I prefer to finish it and after merge we'll see. It seems simpler to do the stacking as proposed by @SoScared - to display each unit type once (like in other mods) with number.

**Comment by GraionDilach** [comment]
You already did a 'major' UI change by upsizing the cameos. I also don't see how a second line is harder than implementing stacking. Mind you that @SoScared isn't a coder. Stacking will mess because the clock animation is either based on the shadow (TD) either based on some arbitratry alpha values (TS onwards) and you can't come up with a general rule to work with both.

**Comment by rob-v** [comment]
This stacking is needed only for TD, no? There is already time 00:03, so it can be the same color.

**Comment by SoScared** [comment]
@GraionDilach I'm not a coder but using second lines would IMO look ugly in comparison.

**Comment by rob-v** [comment]
PR #13023

**Comment by SoScared** [comment]
Added this comment here as not to spam the PR too much with .png images.

I ran it through some Debug "stress tests" and believe I exposed a few discrepancies:
![ra](https://cloud.githubusercontent.com/assets/2715583/24327126/aaff7b7e-11c0-11e7-918a-8599d351fbc0.png)
![td](https://cloud.githubusercontent.com/assets/2715583/24327127/ad045b6a-11c0-11e7-98ba-9c32796e7889.png)

A 1-pixel glitch on the top right corner (all mods) due to I'm guessing the hidden scroll bar area.

I believe the rows would look a bit more attractive with generally _slightly_ more space between them, e.g. 5 pixels space in-between (instead of 3) horizontally and 7-8 pixel space in-between vertically.

**Comment by SoScared** [comment]
Any point on keeping this issue open? Seem like the full brunt of this issue hit #13023 and with it set the course for any future development with the feature.

**Comment by penev92** [comment]
That PR is closed but not finished, so IMO this should stay open until resolved.
Not like it's a #WONTFIX.

**Comment by SoScared** [comment]
I see #13023 on the future milestone so if ever finished it starts there. Also due to the ambivalent feature development the past 3,5 years I've requested the bounty refund on this issue but from what I'm told it I can only request my own share unless this issue is closed. The other fund backers were in on it to increase the pool.

**Comment by anjew175** [comment]
The scope of this specific issue has been limited to just Red Alert. So I'm curious what is needed to be changed or added to make this work.

I understand that it would be preferable to add a comprehensive and versatile UI that adapts across all mods however, as evident in the discussion, this is an issue with the differences in mods. Perhaps, a catch-all spectator UI is not the right direction, instead mods should embrace more appropriate designs.

**Comment by pchote** [comment]
@penev92 was able to resolve those problems and prototype a nicer looking UI without too much trouble, and made a start on the code quality issues.  Unfortunately his own time constraints meant that he was not able to complete that.

**Comment by pchote** [comment]
Also, I can't speak for the others but i'm certainly not happy with the idea of having to maintain multiple sets of observer code that are in different states of disrepair.
